hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableRepresentationStrategyPojo.java

```diff
@@ -83,39 +117,75 @@ private static <T> JavaType<T> resolveEmbeddableJavaType(
+			RuntimeModelCreationContext creationContext) {
+		if ( bootDescriptor.isPolymorphic() ) {
+			final Collection<String> embeddableClassNames = bootDescriptor.getDiscriminatorValues().values();
+			final Map<String, EmbeddableInstantiator> result = new HashMap<>( embeddableClassNames.size() );
+			for ( final String embeddableClassName : embeddableClassNames ) {
+				result.put( embeddableClassName.intern(), determineInstantiator(
+						bootDescriptor,
+						subclassesByName.get( embeddableClassName ),
+						reflectionOptimizer,
+						runtimeDescriptorAccess,
+						creationContext
+				) );
+			}
+			return Collections.unmodifiableMap( result );
+		}
+		else {
+			return Map.of(
+					bootDescriptor.getComponentClassName().intern(),
+					customInstantiator != null ?
+							customInstantiator :
+							determineInstantiator(
+									bootDescriptor,
+									bootDescriptor.getComponentClass(),
+									reflectionOptimizer,
+									runtimeDescriptorAccess,
+									creationContext
+							)
+			);
+		}
+	}
+
+	private static EmbeddableInstantiator determineInstantiator(
+			Component bootDescriptor,
+			Class<?> embeddableClass,
+			ReflectionOptimizer reflectionOptimizer,
 			Supplier<EmbeddableMappingType> runtimeDescriptorAccess,
 			RuntimeModelCreationContext creationContext) {
 		if ( reflectionOptimizer != null && reflectionOptimizer.getInstantiationOptimizer() != null ) {
 			final ReflectionOptimizer.InstantiationOptimizer instantiationOptimizer = reflectionOptimizer.getInstantiationOptimizer();
 			return new EmbeddableInstantiatorPojoOptimized(
-					getEmbeddableJavaType(),
+					embeddableClass,
 					runtimeDescriptorAccess,
 					instantiationOptimizer
 			);
 		}
 
-		if ( bootDescriptor.isEmbedded() && ReflectHelper.isAbstractClass( bootDescriptor.getComponentClass() ) ) {
+		if ( bootDescriptor.isEmbedded() && ReflectHelper.isAbstractClass( embeddableClass ) ) {
 			return new EmbeddableInstantiatorProxied(
-					bootDescriptor.getComponentClass(),
+					embeddableClass,
 					runtimeDescriptorAccess,
 					creationContext.getServiceRegistry()
 							.requireService( ProxyFactoryFactory.class )
-							.buildBasicProxyFactory( bootDescriptor.getComponentClass() )
+							.buildBasicProxyFactory( embeddableClass )
 			);
 		}
 
-		return new EmbeddableInstantiatorPojoStandard( getEmbeddableJavaType(), runtimeDescriptorAccess );
-	}
-
-	@Override
-	public ReflectionOptimizer getReflectionOptimizer() {
-		return reflectionOptimizer;
+		return new EmbeddableInstantiatorPojoStandard( embeddableClass, runtimeDescriptorAccess );
 	}
 
-	@Override
-	protected PropertyAccess buildPropertyAccess(Property bootAttributeDescriptor) {
+	private PropertyAccess buildPropertyAccess(
+			Property bootAttributeDescriptor,
+			Class<?> embeddableClass,
+			boolean requireSetters) {
 		PropertyAccessStrategy strategy = bootAttributeDescriptor.getPropertyAccessStrategy( getEmbeddableJavaType().getJavaTypeClass() );
```

Comment:
Maybe this is wrong?

```suggestion
		PropertyAccessStrategy strategy = bootAttributeDescriptor.getPropertyAccessStrategy( embeddableClass );
```